### PR TITLE
Update/5348 trigger ces on changing setting

### LINKED ISSUE #5348: Customer effort score - Trigger on changing a setting
Part of #5259

Trigger the customer effort score modal when changing a setting - on the existing `wcadmin_settings_change` tracks event.

Copy should be `How easy was it to update your settings?`

### PR DESCRIPTION
**Pull request by moon0326**
Fixes #5348 

This PR triggers the CES survey when settings get updated.

- Added a base unit test class for the upcoming changes.

### Screenshots
Screen Shot 2020-11-09 at 10 21 19 AMth_

### Detailed test instructions:

1. Navigate to WooCommerce -> Settings
2. Change a value
3. Click [ Save Changes ]
4. You should see the snackbar at the bottom left corner on the next page.

To repeat the test, delete `woocommerce_ces_shown_for_actions` and `woocommerce_ces_tracks_queue` from the `wp_options` table.

### CODE REVIEW

File: tests/features/class-wc-tests-ces-tracks.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
+<?php
+/**
+ * Onboarding Themes Tests.
+ *
+ * @package WooCommerce\Admin\Tests\OnboardingThemes
```
Comment (mattsherman) [inline tests/features/class-wc-tests-ces-tracks.php:5]: Comments need to be updated! :-)

Comment (moon0326) [reply tests/features/class-wc-tests-ces-tracks.php]: @mattsherman Thank you for the catch :) Updated it with https://github.com/woocommerce/woocommerce-admin/pull/5569/commits/103ff9682d7074fe0b3e25c33b738794c4d97025

**Comment by moon0326** [comment]
Looking into the failed test.

File: src/Features/CustomerEffortScoreTracks.php [hunk 2]
```diff
@@ -40,17 +40,44 @@ class CustomerEffortScoreTracks {
...
 	public function __construct() {
+		add_action(
+			'admin_init',
+			array(
+				$this,
+				'maybe_clear_ces_tracks_queue',
+			)
+		);
+
+		$this->add_actions();
```
Comment (mattsherman) [inline src/Features/CustomerEffortScoreTracks.php:68]: Is there a reason why we wouldn't want 'admin_init' to also be setup in `add_actions`? And in particular, only is tracking is allowed?

If we do want to keep this separate, maybe `add_actions` should be renamed something like `maybe_enable_survey_enqueing` or `enable_survey_enqueing_if_tracking_is_enabled`.

As it stands, it feels a little confusing/unorganized that we add an action handler and then call another function that says it too is adding actions.

File: src/Features/CustomerEffortScoreTracks.php [hunk 3]
```diff
@@ -40,17 +40,44 @@ class CustomerEffortScoreTracks {
...
+	}
+
+	/**
+	 * Add actions that require woocommerce_allow_tracking.
+	 */
+	private function add_actions() {
+		// Only enqueue a survey if tracking is allowed.
+		$allow_tracking = 'yes' === get_option( 'woocommerce_allow_tracking', 'no' );
```
Comment (mattsherman) [inline src/Features/CustomerEffortScoreTracks.php:76]: I like this optimization! No need to set up the action handlers if we aren't ever going to do anything in them. Nice!

File: src/Features/CustomerEffortScoreTracks.php [hunk 4]
```diff
@@ -242,4 +266,26 @@ function( $item ) use ( $clear_ces_tracks_queue_for_page ) {
...
+
+		$this->enqueue_to_ces_tracks(
+			array(
+				'action'    => self::SETTINGS_CHANGE_ACTION_NAME,
+				'label'     => __(
+					'How easy was it to update your settings?',
+					'woocommerce-admin'
+				),
+				'pagenow'   => 'woocommerce_page_wc-settings',
+				'adminpage' => 'woocommerce_page_wc-settings',
```
Comment (mattsherman) [inline src/Features/CustomerEffortScoreTracks.php:286]: Hmm... do we have anywhere where we document how `pagenow` and `adminpage` should be used? I sorta just copy-pasted my settings in #5540, but seeing how you have set these, it makes me question

(a) what they should be set to, and 
(b) if we really need both `pagenow` and `adminpage`

/cc @becdetat

File: tests/features/class-wc-tests-ces-tracks.php [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+ *
+ * @package Automattic\WooCommerce\Admin\Features
+ */
+
+use \Automattic\WooCommerce\Admin\Features\CustomerEffortScoreTracks;
+
+/**
+ * Class WC_Tests_CES_Tracks
+ */
+class WC_Tests_CES_Tracks extends WC_Unit_Test_Case {
```
Comment (mattsherman) [inline tests/features/class-wc-tests-ces-tracks.php:13]: Thanks for including some tests!

**Review by mattsherman — CHANGES_REQUESTED** [review]
This looks like a great start. I left a few minor comments.

Works as expected.

File: src/Features/CustomerEffortScoreTracks.php (hunk 2 above)
Comment (moon0326) [reply src/Features/CustomerEffortScoreTracks.php:68]: I think we can include admin_init in the add_actions as well. I assumed that we need to run maybe_clear_ces_tracks_queue on every call, but it looks like that's not the case after looking at the code again. In that case, I think we can remove add_actions and move the code back to the constructor. I totally agree that add_actions is not the best name for the method.

File: src/Features/CustomerEffortScoreTracks.php (hunk 4 above)
Comment (moon0326) [reply src/Features/CustomerEffortScoreTracks.php:286]: Looking at the https://github.com/woocommerce/woocommerce-admin/blob/update/5348-trigger-ces-on-changing-setting/client/customer-effort-score-tracks/customer-effort-score-tracks-container.js#L34, those two variables are used to filter the queue. The values can be retrieved from `window.pagenow` and `window.adminpage`.

File: tests/features/class-wc-tests-ces-tracks.php (hunk 5 above)
Comment (moon0326) [reply tests/features/class-wc-tests-ces-tracks.php:13]: Sure thing! I will add more tests once we merge other PRs together.

File: src/Features/CustomerEffortScoreTracks.php (hunk 4 above)
Comment (mattsherman) [reply src/Features/CustomerEffortScoreTracks.php:286]: Thanks!

**Review by mattsherman — APPROVED** [review]
Thanks for making the updates @moon0326! This still works as expected.

Nice job on this PR!

**Pull request merged**